iothub/device/samples/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -181,79 +181,47 @@ private async Task OnC2dMessageReceivedAsync(Message message, object context)
         // has a void return type. As a result, any operation within this block will be executed unmonitored on another thread.
         // To prevent multi-threaded synchronization issues, the async method InitializeClientAsync being called in here first grabs a lock before attempting to
         // initialize or dispose the device client instance; the async method GetTwinAndDetectChangesAsync is implemented similarly for the same purpose.
-        private async void ConnectionStatusChangeHandler(ConnectionInfo connectionInfo)
+        private async void ConnectionStatusChangeHandlerAsync(ConnectionInfo connectionInfo)
         {
-            var status = connectionInfo.Status;
-            var reason = connectionInfo.ChangeReason;
-            Console.WriteLine($"Connection status changed: status={status}, reason={reason}");
+            ConnectionStatus status = connectionInfo.Status;
+            ConnectionStatusChangeReason reason = connectionInfo.ChangeReason;
+            Console.WriteLine($"Connection status changed: status={status}, reason={reason}, recommendation {connectionInfo.RecommendedAction}");
+
+            // In our case, we can operate with more than 1 shared access key and attempt to fall back to a secondary.
+            // We'll disregard the SDK's recommendation and attempt to connect with the second one.
+            if (status == ConnectionStatus.Disconnected
+                && reason == ConnectionStatusChangeReason.BadCredential
+                && _deviceConnectionStrings.Count > 1)
+            {
+                // When getting this reason, the current connection string being used is not valid.
+                // If we had a backup, we can try using that.
+                _deviceConnectionStrings.RemoveAt(0);
+                Console.WriteLine($"The current connection string is invalid. Trying another.");
+                await InitializeAndSetupClientAsync(s_cancellationTokenSource.Token);
+                return;
+            }
 
-            switch (status)
+            // Otherwise, we follow the SDK's recommendation.
+            switch (connectionInfo.RecommendedAction)
```

Comment:
This is what @timtay-microsoft was looking for.

Reply:
Indeed